pkg/clusteragent/custommetrics/provider.go

```diff
@@ -34,24 +36,101 @@ type datadogProvider struct {
 	client dynamic.Interface
 	mapper apimeta.RESTMapper
 
-	values          map[provider.CustomMetricInfo]int64
 	externalMetrics []externalMetric
 	resVersion      string
 	store           Store
+	serving         bool
 	timestamp       int64
 	maxAge          int64
 }
 
 // NewDatadogProvider creates a Custom Metrics and External Metrics Provider.
-func NewDatadogProvider(client dynamic.Interface, mapper apimeta.RESTMapper, store Store) provider.MetricsProvider {
+func NewDatadogProvider(ctx context.Context, client dynamic.Interface, mapper apimeta.RESTMapper, store Store) provider.MetricsProvider {
 	maxAge := config.Datadog.GetInt64("external_metrics_provider.local_copy_refresh_rate")
-	return &datadogProvider{
+	d := &datadogProvider{
 		client: client,
 		mapper: mapper,
-		values: make(map[provider.CustomMetricInfo]int64),
 		store:  store,
 		maxAge: maxAge,
 	}
+	go d.externalMetricsSetter(ctx)
+	return d
+}
+
+type timer struct {
+	period time.Duration
+	timer  time.Timer
+}
+
+func (t *timer) resetTimer() {
+	t.timer.Stop()
+	t.timer = *time.NewTimer(t.period)
+}
+
+func createTimer(period time.Duration) *timer {
+	return &timer{period, *time.NewTimer(period)}
+}
+
+func (p *datadogProvider) externalMetricsSetter(ctx context.Context) {
+	log.Infof("Starting async loop to collect External Metrics")
+	tick := time.NewTicker(time.Duration(p.maxAge) * time.Second)
+	defer tick.Stop()
+	out := createTimer(3 * time.Duration(p.maxAge) * time.Second)
+	defer out.timer.Stop()
+
+	// If we exceed 3 retries trying to access the ConfigMap, we permafail and stop trying to refresh the External Metrics.
```

Comment:
Could we have used the `Retrier` here ?

Reply:
A retrier will be implemented in a different PR, as we want to be more resilient if someone deletes the ConfigMap.